include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -357,11 +455,111 @@ __parallel_transform_reduce(_ExecutionPolicy&& __exec, _Up __u, _LRp __brick_lea
             __is_first = false;
         ::std::swap(__offset_1, __offset_2);
         __n = __n_groups;
-        __n_items = (__n - 1) / __iters_per_work_item + 1;
-        __n_groups = (__n - 1) / __size_per_work_group + 1;
+        __n_items = __ceiling_div(__n, __iters_per_work_item);
+        __n_groups = __ceiling_div(__n, __size_per_work_group);
     } while (__n > 1);
 
-    return __future(__reduce_event, sycl::buffer(__temp, sycl::id<1>(__offset_2), sycl::range<1>(1)));
+    return __future(__reduce_event, __res);
+}
+
+// General version of parallel_transform_reduce. Calls optimized kernels.
+template <typename _Tp, typename _ReduceOp, typename _TransformOp, typename _Functor, typename _ExecutionPolicy,
+          typename _InitType, oneapi::dpl::__internal::__enable_if_device_execution_policy<_ExecutionPolicy, int> = 0,
+          typename... _Ranges>
+auto
+__parallel_transform_reduce(_ExecutionPolicy&& __exec, _ReduceOp __reduce_op, _TransformOp __transform_op,
+                            _InitType __init, _Ranges&&... __rngs)
```

Comment:
Can we come up with a better name for `_Functor` here?  
`_TransformOp2` or `_Transform2ndPassOp` or something?

Is it always (and will it always be) a no-op functor?

Reply:
Good point. I think we will always use a no-op functor since the transform operator is applied on the first pass only. I moved `_Functor` to the inner `__parallel_transform_reduce` pattern.